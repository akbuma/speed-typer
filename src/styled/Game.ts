import styled from "styled-components";

export const StyledGame = styled.div`
  height: 75vh;
  max-height: 500px;
  display: grid;
  grid-template-rows: 50px 1fr;
  grid-template-columns: minmax(50px, auto) 1fr minmax(50px, auto);
`;

export const StyledScore = styled.p`
  font-size: 1.5rem;
`;

export const StyledTimer = styled.p`
  font-size: 1.5rem;
  grid-column: 3;
`;

export const StyledCharacterGame = styled.p`
  font-size: 15rem;
  grid-row: 2;
  grid-column: 1/4;
  text-align: center;
  color: var(--accent-color);
  margin-top: 15vh;
`;

export const StyledCharacterGameOver = styled.p`
  font-size: 15rem;
  grid-row: 2;
  grid-column: 1/4;
  text-align: center;
  color: var(--accent-color);
`;

export const StyledDisplay = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  height: 65vh;
`;

export const StyledOptions = styled.div`
  margin: 0px auto;
  margin-bottom: 15px;
  width: 200px;
`;
