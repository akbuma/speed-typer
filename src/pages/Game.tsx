import React, { useState, useEffect, useCallback } from "react";
import { History, LocationState } from "history";

import {
  StyledGame,
  StyledScore,
  StyledTimer,
  StyledCharacterGame,
  StyledDisplay,
} from "../styled/Game";
import { Strong } from "../styled/Random";
import { useScore } from "../contexts/ScoreContext";

interface GameProps {
  history: History<LocationState>;
}

export default function Game({ history }: GameProps) {
  const { score, setScore } = useScore();

  const MAX_SECONDS = 20;
  const characters = "abcdefghijklmnopqrstuvwxyz0123456789";

  const [currentCharacter, setCurrentCharacter] = useState("");
  const [ms, setMs] = useState("0");
  const [seconds, setSeconds] = useState(MAX_SECONDS.toString());

  useEffect(() => {
    setRandomCharacter();
    setScore(0);
    const currentTime = new Date();
    const interval = setInterval(() => updateTime(currentTime), 1);
    return () => clearInterval(interval);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (parseInt(seconds) <= -1) {
      history.push("/gameOver");
    }
  }, [seconds, ms, history]);

  // updates whenever currentCharacter is updated, otherwise currentCharacter will always equal the initial state in the function
  const keyUpHandler = useCallback(
    // eslint-disable-next-line react-hooks/exhaustive-deps
    (e: KeyboardEvent) => {
      if (e.key === currentCharacter) {
        setScore(score + 1);
      } else {
        if (score > 0) {
          setScore(score - 1);
        }
      }
      setRandomCharacter();
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [currentCharacter]
  );

  const setRandomCharacter = () => {
    const randomInt: number = Math.floor(Math.random() * 36);
    setCurrentCharacter(characters[randomInt]);
  };

  useEffect(() => {
    document.addEventListener("keyup", keyUpHandler);
    // remove the event listener when the component re-renders
    return () => {
      document.removeEventListener("keyup", keyUpHandler);
    };
  }, [keyUpHandler]);

  const updateTime = (startTime: Date) => {
    const endTime = new Date();
    const msPassedStr: number = endTime.getTime() - startTime.getTime();
    const formattedMSString = ("0000" + msPassedStr).slice(-5);
    const updatedSeconds =
      MAX_SECONDS - parseInt(formattedMSString.substring(0, 2)) - 1;
    const updatedMs = (
      1000 - parseInt(formattedMSString.substring(formattedMSString.length - 3))
    )
      .toString()
      .slice(0, 2);
    setSeconds(updatedSeconds.toString().padStart(2, "0"));
    setMs(updatedMs.toString().padStart(2, "0"));
  };

  return (
    <StyledGame>
      <StyledScore>
        Score: <Strong>{score}</Strong>
      </StyledScore>
      <StyledCharacterGame>
        {currentCharacter.toUpperCase()}
      </StyledCharacterGame>
      <StyledTimer>
        Time:{" "}
        <Strong>
          {seconds}:{ms}
        </Strong>
      </StyledTimer>
    </StyledGame>
  );
}
