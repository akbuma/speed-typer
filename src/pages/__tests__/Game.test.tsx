import React from "react";
import renderer from "react-test-renderer";
import Game from "../Game";

describe("Game Class", () => {
  it("renders correct styles", () => {
    //@ts-ignore
    const tree = renderer.create(<Game />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
