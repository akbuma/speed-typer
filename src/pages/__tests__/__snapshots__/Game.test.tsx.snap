// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game Class renders correct styles 1`] = `
<div
  className="sc-bdfBwQ cPnbuW"
>
  <p
    className="sc-gsTCUz cUGGCi"
  >
    Score: 
    <strong
      className="sc-gKsewC jqTwMs"
    >
      0
    </strong>
  </p>
  <p
    className="sc-hKgILt hgEJIr"
  >
    A
  </p>
  <p
    className="sc-dlfnbm hhKTiW"
  >
    Time:
     
    <strong
      className="sc-gKsewC jqTwMs"
    >
      5
      :
      0
    </strong>
  </p>
</div>
`;
